Configure react apps services

spaceo opened the pull request:
#### Link to issue

https://reload.atlassian.net/browse/DDFSOEG-324

#### Description

Since all react app services has be configured from outside we have to define them.
Furthermore we need to define http domains for the CI setup.

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

This is part II of: 
https://github.com/danskernesdigitalebibliotek/dpl-cms/pull/181

spaceo committed:
Secure CI variable type in dev:start task

In some Task versions we got an error when checking the value of the CI
environment variable if it was not set

spaceo committed:
First version of pa11y ci setup

spaceo committed:
Search mappings in place

spaceo committed:
Add search path to pa11y conf

spaceo committed:
Correcting url pattern matching

spaceo committed:
Ready for common

spaceo committed:
We'll try to compile on runtime

That's more appropriate

spaceo committed:
Move config to the isolated pa11y dir

Since we have everything in one directory there is no need to disturb
the root project with that

spaceo committed:
Move dependencies inside pa11y

To make it self contained

spaceo committed:
Get mappings structure in order

To simplify the setup we move all mapping definitions inside a mappings
folder with sub folders for the different areas

spaceo committed:
Clean up scripts

spaceo committed:
Add mapping for dpl-cms branch fetching

spaceo committed:
Add wiremock env and right yarn cmds to pa11y task

spaceo committed:
Add develop dependencies dpl-cms/dpl-design-system

We need latest version of dpl packages

spaceo committed:
Make sure wiremock container is ready

before starting node-chrome container

kasperg committed:
Use non-https version of FBS hostname during testing

It seems as if Wiremock has problems intercepting HTTPS requests.

Our preferred way to do this is via the configuration management system.
As the hostname is not yet configurable via the UI we have to guess 
regarding that the future hostname will be.

Remember to also set the CI environment variable in the CLI container
to make it equivalent to the PHP container.

spaceo committed:
Make sure that json data is an array

because it IS an array!

spaceo committed:
Make sure that query regex is wiremock compatible

The test tool in the GUI is lying.
But the wiremock log never lies

spaceo committed:
Corrections to mappings for search and work page

My eyes opened after I got web proxying in chrome to work

spaceo committed:
Add work page test to Pa11y

spaceo committed:
Pass wiremock proxy properly to pa11y cmd

spaceo committed:
Try only to run tsc inside pa11y dir

spaceo committed:
Try only to run tsc inside pa11y dir II

spaceo committed:
Try only to run tsc inside pa11y dir III

spaceo committed:
Add config manager class property

spaceo committed:
Try to fix discrepency between composer.* files

spaceo committed:
Make pa11y conform to same setup as cypress

* Let .pa11y stay in root of repo like othe conf files
* Use only package.json in root. Deps only defined once
* Make scripts in package.json and task cmd aware of new reality

spaceo committed:
Remove deprecated node_modules

A leftover from having a package.json locally.
The idea was skipped again.

spaceo committed:
Use HTTP_PROXY env in in wiremock mapping

Why not. It is already there...

spaceo committed:
Document the usage of picsum photos

in wiremock mappings

spaceo committed:
Document templating of availability service

properties

spaceo committed:
Return faust id in the cover response

Just to make it more real-ish

spaceo committed:
Remove user token mapping

Since the UI does not rely on a working token request
we can leave it out.

spaceo committed:
Remove mapping for campaigns from the cms

Since we in general avoid mappings for the cms.

spaceo committed:
Remove deprecated files

Used in first setup of pa11y ci. We do not need them anymore.

spaceo committed:
Simplyfy pa11y:ci yarn task

No need to duplicate code.

spaceo committed:
Upload pa11y screenshots for CI debugging

spaceo committed:
Make shure that pa11y screenshots dir exists

spaceo committed:
Configure react apps services

Since all react app services has be configured from outside
we have to define them.
Furthermore we need to define http domains for the CI setup.

kasperg commented on assets/all.settings.php:
```diff
@@ -60,6 +69,16 @@
+  // Set service base urls for the react apps.
+  // We need http domains for testing in CI context.
+  $config['dpl_react_apps.services'] = [
+    'fbs' => ['base_url' => 'http://fbs-openplatform.dbc.dk'],
```
Please reconsider this definition.

The current value duplicates what is already available for `$config['dpl_fbs.settings']`. We only need one of them.

The current approach was chosen due to an expectation of a future `dpl_fbs` module as hinted danskernesdigitalebibliotek/dpl-cms#131 and danskernesdigitalebibliotek/dpl-cms#153 and repeated in danskernesdigitalebibliotek/dpl-cms#186. The current approach goes against this by centralizing all service urls in the `dpl_react_apps` module. This is bound to cause conflicts between the two approaches.

Naming might also cause confusion. Configuration is generally in the format `[module_name].settings.foo.bar`. This uses `[module_name].services`. Services has a different meaning in Drupal/Symfony.

spaceo replied:
I thought that there would come a remark like this :)
So do you have a suggestion for consolidating of the fbs ulrs?
Does all the services belong to the dpl_cms profile?

kasperg replied:
The intention for the structure of the FBS url was that it would belong to a future `dpl_fbs` module.  danskernesdigitalebibliotek/dpl-cms#186 intends to add this and a `dpl_publizon` module. This points to a module per service approach.

If we are to follow this approach then all the remaining urls could be places in similar configuration e.g. `dpl_cover_service`, `dpl_fbi`, `dpl_material_list` etc. The current approach shows that there does not have to be a module for the configuration to exist. On the other hand one might argue that it is silly to use configuration names for something that does not existing and we have no plan to create in the foreseeable future.

In any case: The code as is in direct conflict with what already exists for FBS and [what is being implemented for Publizon](https://github.com/danskernesdigitalebibliotek/dpl-cms/pull/186/files#diff-ed2ad4e359fcd058de99be911d88afeaf7a88e96d6a9b45d9974d4976747863a). I suggest that we align the configuration of these two services with their corresponding modules.

For the rest of the services I am personally fine with either keeping them as is or "preparing" for future service-specific modules.

spaceo replied:
I'l l fix

kasperg commented on web/modules/custom/dpl_react_apps/src/Controller/DplReactAppsController.php:
```diff
@@ -345,6 +347,20 @@ public static function dplCmsBaseUrl(): string {
+  /**
+   * Get the base url of the API's exposed by this site.
+   *
+   *  An array of base urls.
```
This looks like leftovers from a missing `@return` statement :smile:

spaceo replied:
Fixed

kasperg reviewed (commented):
There are some conflicting approaches here and also quite a few static analysis messages.

spaceo committed:
Move fbs and publizon url settings responsibilty

Since the base urls are going to be defined by other modules
we need to treat them differently.

kasperg commented on web/modules/custom/dpl_react_apps/src/Controller/DplReactAppsController.php:
```diff
@@ -345,6 +348,38 @@ public static function dplCmsBaseUrl(): string {
+    if (getenv('CI')) {
+      $services['fbs']['base_url'] = 'http://fbs-openplatform.dbc.dk';
+      $services['publizon']['base_url'] = 'http://pubhub-openplatform.test.dbc.dk';
+    }
```
I wonder why you ended up with this solution as it differs from what already is:

1. We try to keep our configuration handling for specific environments like CI in `*.settings.php` like we do for other services
2. `all.settings.php` already has handling for FBS

I thought something might be wrong with my assumptions so I tried to test what I thought would be the right approach for this in reload/dpl-cms#5. As far as I can tell from running `task ci:reset` with these changes it works.

spaceo replied:
We found common ground. Thx for feedback! ;)

spaceo committed:
Prepare base url settings for future dpl modules

Since the coming dpl_fbs and dpl_publizon modules are defining
the service baseurl in the settings like: [MODULE_NAME].settings.base_url
we make sure to prepare for that.

spaceo committed:
Merge branch 'opsaetning-af-a11y-check-i-gh-actions-pa-dpl-cms-DDFSOEG-324' into opsaetning-af-a11y-check-i-gh-actions-pa-dpl-cms-DDFSOEG-324-part-II
